robot/impl/resource_manager.go

```diff
@@ -537,115 +541,167 @@ func (manager *resourceManager) completeConfig(
 		manager.logger.CDebugw(ctx, "error resolving dependencies", "error", err)
 	}
 
-	resourceNames := manager.resources.ReverseTopologicalSort()
+	// sort resources into topological "levels" based on their dependencies. resources in
+	// any given level only depend on resources in prior levels. this makes it safe to
+	// process resources within a level concurrently as long as levels are processed in
+	// order.
+	levels := manager.resources.ReverseTopologicalSortInLevels()
 	timeout := rutils.GetResourceConfigurationTimeout(manager.logger)
-	for _, resName := range resourceNames {
-		select {
-		case <-ctx.Done():
-			return
-		default:
-		}
-
-		resChan := make(chan struct{}, 1)
-		resName := resName
-		ctxWithTimeout, timeoutCancel := context.WithTimeout(ctx, timeout)
-		defer timeoutCancel()
-
-		cleanup := rutils.SlowStartupLogger(
-			ctx, "Waiting for resource to complete (re)configuration", "resource", resName.String(), manager.logger)
-
-		lr.reconfigureWorkers.Add(1)
-		goutils.PanicCapturingGo(func() {
-			defer func() {
-				cleanup()
-				resChan <- struct{}{}
-				lr.reconfigureWorkers.Done()
-			}()
-			gNode, ok := manager.resources.Node(resName)
-			if !ok || !gNode.NeedsReconfigure() {
-				return
-			}
-			if !(resName.API.IsComponent() || resName.API.IsService()) {
+	for level, resourceNames := range levels {
+		// we use an errgroup here instead of a normal waitgroup to conveniently bubble
+		// up errors in resource processing goroutinues that warrant an early exit.
+		var levelErrG errgroup.Group
+		lr.logger.Infof("--- Processing level %d resources ---", level)
+		for _, resName := range resourceNames {
+			select {
+			case <-ctx.Done():
 				return
+			default:
 			}
 
-			var verb string
-			conf := gNode.Config()
-			if gNode.IsUninitialized() {
-				verb = "configuring"
-				gNode.InitializeLogger(
-					manager.logger, resName.String(), conf.LogConfiguration.Level,
-				)
-			} else {
-				verb = "reconfiguring"
-			}
-			manager.logger.CInfow(ctx, fmt.Sprintf("Now %s resource", verb), "resource", resName)
+			resName := resName
+			// processResource is intended to be run concurrently for each resource
+			// within a topological sort level. if any processResource function returns a
```

Comment:
processResource only returns an error if context is done right? I think the comment here could be clarified, since processResource errors could imply resource reconfiguration errors, which do not cause `completeConfig` to exit early

Reply:
correct, I'll specify that